### PR TITLE
wxPython: 4.2.0

### PR DESCRIPTION
**Pull request by jamesdbrock**
Add wxPython v4.2.0
Disable wxPython v4.1.1 for Python v3.10, because https://github.com/wxWidgets/Phoenix/issues/2016

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tfmoraes** [comment]
@jamesdbrock I think it's better to use wxgtk32 instead of wxgtk31. wxgtk32 was included in nixpkgs https://github.com/NixOS/nixpkgs/pull/191729

**Comment by jamesdbrock** [comment]
It looks like this PR might be redundant after #190587

That PR uses wxgtk32.

File: pkgs/development/python-modules/wxPython/4.1.nix [hunk 1]
```diff
@@ -32,7 +33,9 @@ in
 buildPythonPackage rec {
   pname = "wxPython";
   version = "4.1.1";
-  disabled = isPy27;
+  # wxPython v4.1 doesn't work with Python v3.10
+  # https://github.com/wxWidgets/Phoenix/issues/2016
+  disabled = isPy27 || isPy310;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wxPython/4.1.nix:38]: Please use pythonOlder/pythonAtLeast instead.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -11868,6 +11868,12 @@ in {
     };
   };
 
+  wxPython_4_2 = callPackage ../development/python-modules/wxPython/4.2.nix {
+    wxGTK = pkgs.wxGTK31.override {
+      withGtk2 = false;
+      withWebKit = true;
+    };
+  };
```
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:11876]: This is really not what we want for python Packages. How can we reduce the duplicated?

File: pkgs/development/python-modules/wxPython/4.2.nix [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+  installPhase = ''
+    ${python.interpreter} setup.py install --skip-build --prefix=$out
+    wrapPythonPrograms
+  '';
+
+  meta = with lib; {
+    broken = stdenv.isDarwin;
+    description = "Cross platform GUI toolkit for Python, Phoenix version";
+    homepage = "http://wxpython.org/";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wxPython/4.2.nix:102]:
```suggestion
    homepage = "https://wxpython.org/";
```

File: pkgs/development/python-modules/wxPython/4.2.nix [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  propagatedBuildInputs = [ pillow numpy attrdict ];
+
+  DOXYGEN = "${doxygen}/bin/doxygen";
+
+  preConfigure = lib.optionalString (!stdenv.isDarwin) ''
+    substituteInPlace wx/lib/wxcairo/wx_pycairo.py \
+      --replace '_dlls = dict()' '_dlls = {k: ctypes.CDLL(v) for k, v in [
+        ("gdk",        "${wxGTK.gtk}/lib/libgtk-x11-3.0.so"),
+        ("pangocairo", "${pango.out}/lib/libpangocairo-1.0.so"),
+        ("cairoLib = None", "cairoLib = ctypes.CDLL('${cairo}/lib/libcairo.so')"),
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wxPython/4.2.nix:85]:
```suggestion
        ("gdk",        "${lib.getLib wxGTK.gtk}/lib/libgtk-x11-3.0.so"),
        ("pangocairo", "${lib.getLib pango}/lib/libpangocairo-1.0.so"),
        ("cairoLib = None", "cairoLib = ctypes.CDLL('${lib.getLib cairo}/lib/libcairo.so')"),
```

File: pkgs/development/python-modules/wxPython/4.2.nix [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "663cebc4509d7e5d113518865fe274f77f95434c5d57bc386ed58d65ceed86c7";
+  };
+
+  # https://github.com/NixOS/nixpkgs/issues/75759
+  # https://github.com/wxWidgets/Phoenix/issues/1316
+  doCheck = false;
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wxPython/4.2.nix:47]:
```suggestion
```

File: pkgs/development/python-modules/wxPython/4.2.nix [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  buildPhase = ''
+    ${python.interpreter} build.py -v build_wx dox etg --nodoc sip build_py
+  '';
+
+  installPhase = ''
+    ${python.interpreter} setup.py install --skip-build --prefix=$out
+    wrapPythonPrograms
+  '';
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wxPython/4.2.nix:98]:
```suggestion

  # https://github.com/NixOS/nixpkgs/issues/75759
  # https://github.com/wxWidgets/Phoenix/issues/1316
  doCheck = false;

```

File: pkgs/development/python-modules/wxPython/4.2.nix [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+, xorgproto
+, gst_all_1
+, libglvnd
+, mesa
+, webkitgtk
+, autoPatchelfHook
+}:
+let
+  dynamic-linker = stdenv.cc.bintools.dynamicLinker;
+in
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wxPython/4.2.nix:33]:
```suggestion

```

**Review by SuperSandro2000 — COMMENTED** [review]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

**Comment by jamesdbrock** [comment]
wxPython_4_2 was added in #190587

**Pull request closed without merging**